producer.go

```diff
@@ -344,9 +388,17 @@ func (w *Producer) router() {
 		select {
 		case t := <-w.transactionChan:
 			w.transactions = append(w.transactions, t)
-			err := w.conn.WriteCommand(t.cmd)
+			err := w.conn.WriteCommandWithContext(t.ctx, t.cmd)
 			if err != nil {
 				w.log(LogLevelError, "(%s) sending command - %s", w.conn.String(), err)
+				if err == context.Canceled || err == context.DeadlineExceeded {
+					// keep the connection alive if related to context timeout
+					// need to do some stuff that's in Producer.popTransaction here
+					w.transactions = w.transactions[1:]
+					t.Error = err
```

Comment:
I don't love this code we've copied in to this context-specific edge case, can't we consolidate this in to `popTransaction`? Do we need to define a context-specific error code for the protocol?